File: skypydb/security/mixins/validation/validate_inputs.py
"""Validation helpers for names, values, filters, and config dictionaries."""

from typing import Any, Dict, Optional
from skypydb.errors import ValidationError
from skypydb.security.mixins.validation.sanitize_values import SanitizeValuesMixin
from skypydb.security.mixins.validation.sql_injection_check import SQLInjectionCheckMixin
from skypydb.security.constants import (
    COLUMN_NAME_PATTERN,
    MAX_COLUMN_NAME_LENGTH,
    MAX_STRING_LENGTH,
    MAX_TABLE_NAME_LENGTH,
    TABLE_NAME_PATTERN,
)


class ValidateInputsMixin(SQLInjectionCheckMixin, SanitizeValuesMixin):
    """Validate user-facing identifiers and value payloads."""

    @classmethod
    def validate_table_name(cls, table_name: str) -> str:
        """Validate table/collection names used to build SQL identifiers."""

        if not table_name:
            raise ValidationError("Table name cannot be empty")
        if not isinstance(table_name, str):
            raise ValidationError("Table name must be a string")
        if len(table_name) > MAX_TABLE_NAME_LENGTH:
            raise ValidationError(
                f"Table name too long (max {MAX_TABLE_NAME_LENGTH} characters)"
            )
        if not TABLE_NAME_PATTERN.match(table_name):
            raise ValidationError(
                "Table name must start with a letter or underscore and contain only "
                "alphanumeric characters, underscores, and hyphens"
            )
        if cls._contains_sql_injection(table_name):
            raise ValidationError("Table name contains potentially dangerous characters")
        return table_name

    @classmethod
    def validate_column_name(cls, column_name: str) -> str:
        """Validate SQL column names used by filtering and schema-like payloads."""

        if not column_name:
            raise ValidationError("Column name cannot be empty")
        if not isinstance(column_name, str):
            raise ValidationError("Column name must be a string")
        if len(column_name) > MAX_COLUMN_NAME_LENGTH:
            raise ValidationError(
                f"Column name too long (max {MAX_COLUMN_NAME_LENGTH} characters)"
            )
        if not COLUMN_NAME_PATTERN.match(column_name):
            raise ValidationError(
                "Column name must start with a letter or underscore and contain only "
                "alphanumeric characters and underscores"
            )
        if cls._contains_sql_injection(column_name):
            raise ValidationError("Column name contains potentially dangerous characters")
        return column_name

    @classmethod
    def validate_string_value(cls, value: str, max_length: Optional[int] = None) -> str:
        """Validate string length constraints for text payload fields."""

        if not isinstance(value, str):
            raise ValidationError("Value must be a string")

        max_len = max_length or MAX_STRING_LENGTH
        if len(value) > max_len:
            raise ValidationError(f"String value too long (max {max_len} characters)")
        return value

    @classmethod
    def validate_data_dict(cls, data: Dict[str, Any]) -> Dict[str, Any]:
        """Validate a data dictionary and sanitize unsafe string values."""

        if not isinstance(data, dict):
            raise ValidationError("Data must be a dictionary")

        validated_data = {}
        for key, value in data.items():
            validated_key = cls.validate_column_name(key)
            if isinstance(value, str):
                validated_value = cls.sanitize_string(value)
            elif isinstance(value, (int, float, bool)):
                validated_value = value
            elif value is None:
                validated_value = None
            else:
                validated_value = cls.sanitize_string(str(value))

            validated_data[validated_key] = validated_value
        return validated_data

    @classmethod
    def validate_filter_dict(cls, filters: Dict[str, Any]) -> Dict[str, Any]:
        """Validate filter dictionaries used by get/query/delete operations."""

        if not isinstance(filters, dict):
            raise ValidationError("Filters must be a dictionary")

        validated_filters = {}
        for key, value in filters.items():
            validated_key = cls.validate_column_name(key)
            if isinstance(value, list):
                validated_value = [
                    cls.sanitize_string(str(v))
                    if not isinstance(v, (int, float, bool, type(None)))
                    else v
                    for v in value
                ]
            elif isinstance(value, str):
                validated_value = cls.sanitize_string(value)
            elif isinstance(value, (int, float, bool)):
                validated_value = value
            elif value is None:
                validated_value = None
            else:
                validated_value = cls.sanitize_string(str(value))

            validated_filters[validated_key] = validated_value
        return validated_filters

    @classmethod
    def validate_config(cls, config: Dict[str, Any]) -> Dict[str, Any]:
        """Validate generic config dictionaries containing table/column type mappings."""

        if not isinstance(config, dict):
            raise ValidationError("Configuration must be a dictionary")

        validated_config = {}
        for table_name, table_config in config.items():
            validated_table_name = cls.validate_table_name(table_name)
            if not isinstance(table_config, dict):
                raise ValidationError(
                    f"Configuration for table '{table_name}' must be a dictionary"
                )
            validated_table_config = {}
            for column_name, column_type in table_config.items():
                validated_column_name = cls.validate_column_name(column_name)
                valid_types = [str, int, float, bool, "str", "int", "float", "bool", "auto"]
                if column_type not in valid_types:
                    raise ValidationError(
                        f"Invalid type for column '{column_name}': {column_type}. "
                        f"Valid types are: {valid_types}"
                    )
                validated_table_config[validated_column_name] = column_type

            validated_config[validated_table_name] = validated_table_config
        return validated_config

def validate_table_name(table_name: str) -> str:
    """Convenience wrapper for validating table/collection names."""

    return ValidateInputsMixin.validate_table_name(table_name)

def validate_column_name(column_name: str) -> str:
    """Convenience wrapper for validating column names."""

    return ValidateInputsMixin.validate_column_name(column_name)

File: skypydb-js/src/security/internal/validation/validate_inputs.ts
import { ValidationError } from "../../../errors";
import {
  COLUMN_NAME_PATTERN,
  MAX_COLUMN_NAME_LENGTH,
  MAX_STRING_LENGTH,
  MAX_TABLE_NAME_LENGTH,
  TABLE_NAME_PATTERN,
} from "../../../security/constants";
import { SanitizeValuesMixin } from "./sanitize_values";
import { SQLInjectionCheckMixin } from "./sql_injection_check";

export class ValidateInputsMixin extends SQLInjectionCheckMixin {
  static validate_table_name(table_name: string): string {
    if (!table_name) {
      throw new ValidationError("Table name cannot be empty");
    }
    if (typeof table_name !== "string") {
      throw new ValidationError("Table name must be a string");
    }
    if (table_name.length > MAX_TABLE_NAME_LENGTH) {
      throw new ValidationError(
        `Table name too long (max ${MAX_TABLE_NAME_LENGTH} characters)`,
      );
    }
    if (!TABLE_NAME_PATTERN.test(table_name)) {
      throw new ValidationError(
        "Table name must start with a letter or underscore and contain only alphanumeric characters, underscores, and hyphens",
      );
    }
    if (this._contains_sql_injection(table_name)) {
      throw new ValidationError(
        "Table name contains potentially dangerous characters",
      );
    }
    return table_name;
  }

  static validate_column_name(column_name: string): string {
    if (!column_name) {
      throw new ValidationError("Column name cannot be empty");
    }
    if (typeof column_name !== "string") {
      throw new ValidationError("Column name must be a string");
    }
    if (column_name.length > MAX_COLUMN_NAME_LENGTH) {
      throw new ValidationError(
        `Column name too long (max ${MAX_COLUMN_NAME_LENGTH} characters)`,
      );
    }
    if (!COLUMN_NAME_PATTERN.test(column_name)) {
      throw new ValidationError(
        "Column name must start with a letter or underscore and contain only alphanumeric characters and underscores",
      );
    }
    if (this._contains_sql_injection(column_name)) {
      throw new ValidationError(
        "Column name contains potentially dangerous characters",
      );
    }
    return column_name;
  }

  static validate_string_value(value: string, max_length?: number): string {
    if (typeof value !== "string") {
      throw new ValidationError("Value must be a string");
    }
    const max_len = max_length ?? MAX_STRING_LENGTH;
    if (value.length > max_len) {
      throw new ValidationError(
        `String value too long (max ${max_len} characters)`,
      );
    }
    return value;
  }

  static validate_data_dict(
    data: Record<string, unknown>,
  ): Record<string, unknown> {
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
      throw new ValidationError("Data must be a dictionary");
    }

    const validated: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      const valid_key = this.validate_column_name(key);
      if (typeof value === "string") {
        validated[valid_key] = SanitizeValuesMixin.sanitize_string(value);
      } else if (
        typeof value === "number" ||
        typeof value === "boolean" ||
        value === null
      ) {
        validated[valid_key] = value;
      } else {
        validated[valid_key] = SanitizeValuesMixin.sanitize_string(
          String(value),
        );
      }
    }
    return validated;
  }

  static validate_filter_dict(
    filters: Record<string, unknown>,
  ): Record<string, unknown> {
    if (
      typeof filters !== "object" ||
      filters === null ||
      Array.isArray(filters)
    ) {
      throw new ValidationError("Filters must be a dictionary");
    }

    const validated: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(filters)) {
      const valid_key = this.validate_column_name(key);
      if (Array.isArray(value)) {
        validated[valid_key] = value.map((item) => {
          if (
            typeof item === "number" ||
            typeof item === "boolean" ||
            item === null
          ) {
            return item;
          }
          return SanitizeValuesMixin.sanitize_string(String(item));
        });
      } else if (typeof value === "string") {
        validated[valid_key] = SanitizeValuesMixin.sanitize_string(value);
      } else if (
        typeof value === "number" ||
        typeof value === "boolean" ||
        value === null
      ) {
        validated[valid_key] = value;
      } else {
        validated[valid_key] = SanitizeValuesMixin.sanitize_string(
          String(value),
        );
      }
    }
    return validated;
  }

  static validate_config(
    config: Record<string, unknown>,
  ): Record<string, unknown> {
    if (
      typeof config !== "object" ||
      config === null ||
      Array.isArray(config)
    ) {
      throw new ValidationError("Configuration must be a dictionary");
    }

    const validated: Record<string, unknown> = {};
    for (const [table_name, table_config] of Object.entries(config)) {
      const valid_table_name = this.validate_table_name(table_name);
      if (
        typeof table_config !== "object" ||
        table_config === null ||
        Array.isArray(table_config)
      ) {
        throw new ValidationError(
          `Configuration for table '${table_name}' must be a dictionary`,
        );
      }
      const validated_table_config: Record<string, unknown> = {};
      for (const [column_name, column_type] of Object.entries(table_config)) {
        const valid_column_name = this.validate_column_name(column_name);
        const valid_types: Array<unknown> = [
          String,
          Number,
          Boolean,
          "str",
          "int",
          "float",
          "bool",
          "auto",
        ];
        if (!valid_types.includes(column_type)) {
          throw new ValidationError(
            `Invalid type for column '${column_name}': ${String(column_type)}. Valid types are: ${valid_types.join(", ")}`,
          );
        }
        validated_table_config[valid_column_name] = column_type;
      }
      validated[valid_table_name] = validated_table_config;
    }
    return validated;
  }
}

export function validate_table_name(table_name: string): string {
  return ValidateInputsMixin.validate_table_name(table_name);
}

export function validate_column_name(column_name: string): string {
  return ValidateInputsMixin.validate_column_name(column_name);
}
